Engine/Settings.cs

```diff
@@ -106,6 +114,68 @@ public Settings(object settings) : this(settings, null)
         {
         }
 
+        /// <summary>
+        /// Add a configurable rule to the rule arguments
+        /// </summary>
+        public Settings AddRuleArgument(string name, Dictionary<string,object>setting)
+        {
+            Dictionary<string,object> currentSetting;
+            if (RuleArguments.TryGetValue(name, out currentSetting))
+            {
+                throw new ArgumentException(name + " already in settings");
+            }
+            RuleArguments.Add(name, setting);
+            return this;
+        }
+
+        /// <summary>
+        /// Add a configuration element to a rule
+        /// </summary>
+        public Settings AddRuleArgument(string rule, string setting, object value)
+        {
+            // the rule does not exist, that's a problem, first create it and then add the key/value pair
+            Dictionary<string, object> settingDictionary;
+            if ( ! RuleArguments.TryGetValue(rule, out settingDictionary))
```

Comment:
variable declaration can be inlined: `out Dictionary<string, object> settingDictionary`, also in a few lines below